### ISSUE TITLE
Suggestion: no redundant/unused generics

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#4755]
### Rule Suggestion

**Is your rule for a general problem or is it specific to your development style?**
General problem

**What does your suggested rule do?**
```ts
type Component<P> = { p: P };
declare function fn<P, T extends Component<P>>(t: T): P;
```

The generic `P` will always be `unknown` (`{}` prior to 3.5).

This should be rewritten to:

``` ts
declare function fn<T extends Component<any>>(t: T): T['p'];
```

There are many times we know that a generic will always be inferred as `unknown`—whenever the generic is not explicitly used by parameters. In this case, I would like a lint rule to remind me to write the function differently.

Note sometimes this code is desired, e.g. if the generic is designed to be passed explicitly instead of inferred.

**List several examples where your rule could be used**

**Additional context**
Potentially related:
- https://twitter.com/SeaRyanC/status/1118634571564630016
- https://twitter.com/ahejlsberg/status/1118642526817144832
- https://github.com/microsoft/dtslint/blob/d9cef9d8aa9b25e8c625e97d42c8733228084afe/docs/no-unnecessary-generics.md

**Comment by ajafff** [comment]
You could use this rule to do the job: https://github.com/fimbullinter/wotan/blob/master/packages/mimir/docs/no-misused-generics.md

> Note sometimes this code is desired, e.g. if the generic is designed to be passed explicitly instead of inferred.

There's no way to detect that. You'd have to disable the rule for such cases. 
There shouldn't be too many cases anyway. Factory functions or constructors of collections (without a parameter to initialize it) come to mind:

```ts
declare function createMap<T>(): MapLike<T>;
const m = createMap<string>();

// could be rewritten to
declare function createMap(): MapLike<any>;
const m: MapLike<string> = createMap();
```

**Comment by OliverJAsh** [comment]
Amazing, thanks!

**Comment by OliverJAsh** [comment]
Should I keep this open to track progress for adding this to tslint, or do we just delegate to Wotan?
